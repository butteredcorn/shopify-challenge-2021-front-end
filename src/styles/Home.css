.home-container {
    width: 90%;
    max-width: 1200px; 
}

.home-container > * {
    margin: 1rem;
}

.home-header {
    min-height: 10vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    font-size: 2.5rem;
    font-weight: bold;
    margin: 1rem;
  }

.main-section {
    display: flex;
    align-items: flex-start;
}

.main-section > * {
    margin: 1rem;
}

