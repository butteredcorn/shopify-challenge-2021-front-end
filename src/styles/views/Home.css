.home-container {
    width: 90%;
    max-width: 1200px; 
}

.home-main-section {
    display: flex;
    align-items: flex-start;
}

.home-main-section > * {
    margin: 1rem;
    width: 50%;
}

@media only screen and (max-width: 600px) {

    .home-container {
        width: 100%;
        max-width: unset; 
    }

    .home-main-section {
        align-items: unset;
        justify-content: unset;
        flex-direction: column;
    }

    .home-main-section > * {
        width: unset;
    }


}
