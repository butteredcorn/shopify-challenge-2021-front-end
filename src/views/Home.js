import { useEffect, useState } from 'react';
import styled from 'styled-components'
import { useLazyQuery } from "@apollo/client"
import { GET_MOVIES_BY_TITLE } from '../graphql/Queries'


import '../styles/Home.css'

import MainHeader from '../components/MainHeader'
import SearchBar from '../components/SearchBar'
import GetMoviesByTitle from '../components/GetMoviesByTitle'
import NominationsList from '../components/NominationsList'

const HomeContainer = styled.section`
    background-color: ${props => props.theme.bg.primary};
    color: ${props => props.theme.text.tertiary};
    border: 1px solid ${props => props.theme.bg.primary};
`
const MainSection = styled.section`
    background-color: ${props => props.theme.bg.primary};
    color: ${props => props.theme.text.tertiary};
    border: 1px solid ${props => props.theme.bg.primary};
`

const Home = () => {
    const [keyword, setKeyword] = useState("")
    const [getMoviesByTitle, { error, loading, data }] = useLazyQuery(GET_MOVIES_BY_TITLE);    
    const [movies, setMovies] = useState([])
    const [nominations, setNominations] = useState([])

    useEffect(() => {
        if (data && data.searchByTitle && data.searchByTitle.Search) {
            console.log(data.searchByTitle.Search)
            if (data.searchByTitle.Search.length > 0) setMovies(data.searchByTitle.Search)
            //else no matching movies and render empty search result message
        }
    }, [data])

    const handleEnter = (key) => {
        if (key && key === "Enter") getMoviesByTitle({variables: {title: keyword}})
    }

    const addNomination = (nomination)  => {
        if (nominations && nominations.filter(n => n.imdbID === nomination.imdbID).length > 0) return;
        const updatedNominations = [...nominations, nomination]
        setNominations(updatedNominations)
        localStorage.setItem("nominations", JSON.stringify(updatedNominations))
    }

    const removeNomination = (id) => {
        const updatedNominations = nominations.filter(n => n.imdbID !== id)
        setNominations(updatedNominations)
        localStorage.setItem("nominations", JSON.stringify(updatedNominations))
    }

    useEffect(() => {
        const nominations = JSON.parse(localStorage.getItem("nominations"))
        setNominations(nominations)
    }, [])

    // useEffect(() => {
    //     localStorage.setItem("nominations", JSON.stringify(nominations))
    // }, [nominations])

    return (
        <HomeContainer className="home-container">
            <MainHeader/>
            <SearchBar keyword={keyword} setKeyword={setKeyword} onEnter={handleEnter}/>
            <MainSection className="main-section">
                <GetMoviesByTitle keyword={keyword} loading={loading} error={error} data={movies} nominate={addNomination} nominations={nominations}/>
                <NominationsList data={nominations} remove={removeNomination}/>
            </MainSection>
        </HomeContainer>
    );
  };
  export default Home;