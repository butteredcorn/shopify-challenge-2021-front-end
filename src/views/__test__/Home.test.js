import { render, screen, cleanup, fireEvent, getByText } from '@testing-library/react';
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import ThemesProvider from '../../providers/ThemeProvider'
import Home from '../Home';
import {lightTheme, darkTheme} from '../../styles/common/theme.css'

const rgb2hex = (rgb) => `#${rgb.match(/^rgb\((\d+),\s*(\d+),\s*(\d+)\)$/).slice(1).map(n => parseInt(n, 10).toString(16).padStart(2, '0')).join('')}`

afterEach(cleanup);

it('renders Home component with "No Nominations Yet! Nominate some movies to get started." and "Movie results will show here after you search." if prop "nominations" is set to empty array and no search has been conducted.', () => {
  render(<BrowserRouter><ThemesProvider><Home nominations={[]}/></ThemesProvider></BrowserRouter>);
  expect(screen.getByText(/Movie results will show here after you search./i)).toBeInTheDocument();
  expect(screen.getByText(/No Nominations Yet! Nominate some movies to get started./i)).toBeInTheDocument();
});

it('renders Home component without "No Nominations Yet! Nominate some movies to get started." if prop "nominations" is set to an array with one movie object.', () => {
    render(<BrowserRouter><ThemesProvider><Home nominations={[{Title: "Shopify Challenge - Front End Developer!", Year: "2021", imdbID: "the_shoppies"}]}/></ThemesProvider></BrowserRouter>);
    const nonExistantMessage = screen.queryByText(/No Nominations Yet! Nominate some movies to get started./i);
    expect(nonExistantMessage).toBeNull();
});

it('renders the Light Theme backgorund color for home-container', () => {
    render(<BrowserRouter><ThemesProvider><Home nominations={[]}/></ThemesProvider></BrowserRouter>);
    const MyHomeRoots = document.getElementsByClassName("home-container")
    const style = window.getComputedStyle(MyHomeRoots[0])
    expect(rgb2hex(style.backgroundColor)).toBe(lightTheme.bg.primary)
});

it('renders the Dark Theme background color for home-container after Theme toggle is clicked.', () => {
    const {container} = render(<BrowserRouter><ThemesProvider><Home nominations={[]}/></ThemesProvider></BrowserRouter>);
    fireEvent(
        getByText(container, 'light Theme'),
        new MouseEvent('click', {
            bubbles: true,
            cancelable: true
        })
    )
    const MyHomeRoots = document.getElementsByClassName("home-container")
    const style = window.getComputedStyle(MyHomeRoots[0])
    expect(rgb2hex(style.backgroundColor)).toBe(darkTheme.bg.primary)
});

it('matches a snapshot', () => {
  const json = renderer.create(<BrowserRouter><ThemesProvider><Home nominations={[{Title: "Shopify Challenge - Front End Developer!", Year: "2021", imdbID: "the_shoppies"}]}/></ThemesProvider></BrowserRouter>).toJSON();
  expect(json).toMatchSnapshot();
});