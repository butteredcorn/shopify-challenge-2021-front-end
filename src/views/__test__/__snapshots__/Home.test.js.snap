// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a snapshot 1`] = `
<section
  className="sc-iqsfdx fPGZVg home-container"
>
  <section
    className="sc-bdvvaa cuEABP"
  >
    <header
      className="home-header"
    >
      <h1>
        <a
          className="nav-link"
          href="/"
          onClick={[Function]}
        >
          The Shoppies
        </a>
      </h1>
      <p
        className="toggle"
        onClick={[Function]}
      >
        light
         Theme
      </p>
    </header>
  </section>
  <section
    className="sc-gsDJrp cGoqlV search-bar-container"
  >
    <section
      className="sc-dkPtyc hBdwQl search-bar-inner"
    >
      <label
        className="sc-eCImvq iMoTva search-bar-title"
      >
        Movie title
      </label>
      <span
        className="sc-hKwCoD kWunAK search"
      >
        <img
          alt="search icon"
          className="search-icon"
          onClick={[Function]}
          src="https://cdn4.iconfinder.com/data/icons/glyphs/24/icons_search-64.png"
        />
        <input
          className="search-input"
          name="keyword"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value=""
        />
      </span>
    </section>
  </section>
  <section
    className="sc-crHlIS dShgku home-main-section"
  >
    <div
      className="sc-gKckTs imZtZy movies-container"
    >
      <div>
        <h3
          className="sc-iCfLBT dpfegV heading"
        >
          Results
        </h3>
        <p>
          Movie results will show here after you search.
        </p>
      </div>
    </div>
    <div
      className="sc-furvIG lcAKQw nominations-container"
    >
      <h3
        className="sc-pVTma jKyYsN heading"
      >
        Nominations
      </h3>
      <ul
        className="nominations"
      >
        <li
          className="nomination"
        >
          <a
            href="/movie"
            onClick={[Function]}
          >
            Shopify Challenge - Front End Developer!
          </a>
           
          <span>
            (
            2021
            )
          </span>
           
          <span>
            <button
              onClick={[Function]}
            >
              Remove
            </button>
          </span>
        </li>
      </ul>
    </div>
  </section>
</section>
`;
